File: packages/client/src/ulid.ts
/**
 * ULID (Universally Unique Lexicographically Sortable Identifier) generation.
 *
 * 26-character Crockford base32: a 10-character 48-bit millisecond timestamp
 * followed by a 16-character 80-bit random component. Calls within the same
 * millisecond increment the random component (monotonic factory), so ids
 * produced back-to-back stay lexicographically sortable and never collide
 * under normal use. No external dependency; safe in Node and browsers.
 *
 * https://github.com/ulid/spec
 */

const ENCODING = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';
const TIME_LEN = 10;
const RANDOM_LEN = 16;
const RANDOM_BYTES = 10; // 80 bits = 16 base32 chars exactly
const RANDOM_BITS = 80n;
const RANDOM_MAX = (1n << RANDOM_BITS) - 1n;

let lastTime = -1;
let lastRandom = 0n;

function getRandomBytes(length: number): Uint8Array {
  const bytes = new Uint8Array(length);
  const source = (globalThis as { crypto?: Crypto }).crypto;
  if (source && typeof source.getRandomValues === 'function') {
    source.getRandomValues(bytes);
    return bytes;
  }
  for (let i = 0; i < length; i++) bytes[i] = Math.floor(Math.random() * 256);
  return bytes;
}

function randomBigInt(): bigint {
  const bytes = getRandomBytes(RANDOM_BYTES);
  let value = 0n;
  for (const byte of bytes) value = (value << 8n) | BigInt(byte);
  return value;
}

function encodeTime(time: number): string {
  let str = '';
  let t = time;
  for (let i = 0; i < TIME_LEN; i++) {
    const mod = t % 32;
    str = ENCODING.charAt(mod) + str;
    t = Math.floor(t / 32);
  }
  return str;
}

function encodeRandom(value: bigint): string {
  let str = '';
  let v = value;
  for (let i = 0; i < RANDOM_LEN; i++) {
    str = ENCODING.charAt(Number(v & 31n)) + str;
    v >>= 5n;
  }
  return str;
}

/** Generate a new 26-character Crockford-base32 ULID, monotonic within a millisecond. */
export function ulid(): string {
  const time = Date.now();
  let random: bigint;
  if (time === lastTime) {
    random = lastRandom + 1n;
    if (random > RANDOM_MAX) {
      // 80 bits of randomness exhausted inside one millisecond: vanishingly
      // rare, but stay monotonic by borrowing the next millisecond.
      lastTime = time + 1;
      random = randomBigInt();
      lastRandom = random;
      return encodeTime(lastTime) + encodeRandom(random);
    }
  } else {
    random = randomBigInt();
  }
  lastTime = time;
  lastRandom = random;
  return encodeTime(time) + encodeRandom(random);
}

File: clients/python/src/atriarch/tracery/ulid.py
"""ULID (Universally Unique Lexicographically Sortable Identifier) generation.

26-character Crockford base32: a 10-character 48-bit millisecond timestamp
followed by a 16-character 80-bit random component. Calls within the same
millisecond increment the random component (monotonic factory), so ids
produced back-to-back stay lexicographically sortable and never collide
under normal use. Thread-safe. Stdlib only.

https://github.com/ulid/spec
"""

from __future__ import annotations

import os
import threading
import time

_ENCODING = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
_TIME_LEN = 10
_RANDOM_LEN = 16
_RANDOM_BYTES = 10  # 80 bits = 16 base32 chars exactly
_RANDOM_BITS = 80
_RANDOM_MAX = (1 << _RANDOM_BITS) - 1

_lock = threading.Lock()
_last_time = -1
_last_random = 0


def _random_bigint() -> int:
    return int.from_bytes(os.urandom(_RANDOM_BYTES), "big")


def _encode_time(time_ms: int) -> str:
    chars: list[str] = []
    t = time_ms
    for _ in range(_TIME_LEN):
        t, mod = divmod(t, 32)
        chars.append(_ENCODING[mod])
    return "".join(reversed(chars))


def _encode_random(value: int) -> str:
    chars: list[str] = []
    v = value
    for _ in range(_RANDOM_LEN):
        v, mod = divmod(v, 32)
        chars.append(_ENCODING[mod])
    return "".join(reversed(chars))


def ulid() -> str:
    """Generate a new 26-character Crockford-base32 ULID, monotonic within a millisecond."""
    global _last_time, _last_random
    with _lock:
        time_ms = int(time.time() * 1000)
        if time_ms == _last_time:
            random_value = _last_random + 1
            if random_value > _RANDOM_MAX:
                # 80 bits of randomness exhausted inside one millisecond:
                # vanishingly rare, but stay monotonic by borrowing the next ms.
                time_ms += 1
                random_value = _random_bigint()
        else:
            random_value = _random_bigint()
        _last_time = time_ms
        _last_random = random_value
        return _encode_time(time_ms) + _encode_random(random_value)
